First release of cleaner graphing code

SuperZooper3 opened the pull request:
Changes made since the initial release of existing code are:
- Object orient lines and axes
- Restructure data point adding system, now can support any frame rate and the data points will catch up to the real time of the animation
- Parameterize FFmpeg path and other setup 
- Break individual animation settings (axis and lines) into separate config file
- Add option to preview final graph to speed up workflow

Commit:
Perform initial release of existing code

Commit:
Make the graph not cringe

SuperZooper3 committed:
Being object orient refactor, in working state

SuperZooper3 committed:
First test for LCF1, outline of features to add

SuperZooper3 committed:
Create config files, oop axies, config spots

SuperZooper3 committed:
Delete config_setup.py

SuperZooper3 committed:
Ignore config_setup.py, only keep example

SuperZooper3 committed:
Clean up code layout and clarify flow

SuperZooper3 committed:
Move all config out of animator script

SuperZooper3 committed:
Add progress bar, set example config

SuperZooper3 committed:
Add filter strengh optioin, fix out of range

SuperZooper3 committed:
document setup, config and usage

SuperZooper3 committed:
add primary axis config validation

Artem-Sotnikov commented:
Having some example data for users will be a good idea. Also, some more descriptive error comments should be present for the scenario where the user does not make the correct copies of config_setup and config_ani

Artem-Sotnikov commented:
I'll take a closer look at the code later, these are some preliminary suggestions

SuperZooper3 commented:
> Having some example data for users will be a good idea. 

I've updated the example data in the /example directory, and its associated config is what is kept as the `config_ani_example`.

I'll add errors to warn when they haven't made copies correctly.

SuperZooper3 committed:
Error catching for missing config

SuperZooper3 committed:
auto-pep8 and clean up some comments

Commit:
Clean up minor things, add troubleshooting to readme

Clean up line lengths in document to make it more consistent, add function to allow graphical offsets of data. Create troubleshooting section regarding font issues (a really annoying issue that I ran into while trying to run the code)

Artem-Sotnikov commented:
I have managed to run and play around with the code. It's a very solid data visualization package, with lots of great features. After you take a look at my changes, I think we're good to merge! Some things to do for future generations may be to set up an option to run this with a venv for more consistent usage long-term.

KavinSatheeskumar commented on data_animation.py:
```diff
@@ -0,0 +1,202 @@
+SKIP_TO_FINAL = config_ani.SKIP_TO_FINAL
+PREVIEW = config_ani.PREVIEW
+
+if SKIP_TO_FINAL:
```
Just a suggestion on code style, but there is a quote from the late, great, Jacob Deery

"Try and have as much of the code as possible in functions in other files, the main file should be as possible."

I think alot of the logic in this file is better placed in functions outside of the file, and potentially make a file called "main.py" which just imports and runs everything

Artem-Sotnikov replied:
That's actually a good point. We should wrap all the code in a function and run it inside of a main.py

SuperZooper3 replied:
Yep I agree. It's currently done like this because it's an evolution of a standalone script, but I'll do that now.

SuperZooper3 replied:
I've given a shot at splitting it up into different files, but given the way this is laid out and how almost everything is interacting with matplotlib plt directly, or is just already importing lots of settings from other files, I think trying to split things out would be significantly more refactoring than I want to do right now.

SuperZooper3 replied:
It's been broken into its own main.py script in [c928684](https://github.com/waterloo-rocketry/data-video-visualization/pull/1/commits/c9286849fd8801e7a57d8e0b004901f1a6766863)

KavinSatheeskumar commented on config_setup_example.py:
```diff
@@ -0,0 +1,5 @@
+## SET YOUR VALUES IN A COPY OF THIS CALLED config_setup.py
```
Is there any reason why you have 2 config files instead of 1?

Artem-Sotnikov replied:
Separating animation setup and overall config makes sense in my opinion

SuperZooper3 replied:
Like Artem says, yeah, one is an overall config (one off) and an animation config (one par animation). The idea is that you just as you make each animation make tons of copies of the animation config one and whichever one is named config_ani is the config that will be used.

KavinSatheeskumar commented on config_ani_example.py:
```diff
@@ -0,0 +1,86 @@
+## SET YOUR SETTINGS IN A COPY OF THIS CALLED config_ani.py
```
Just out of curiosity, is there any reason why you opted to use just a python file, and not something a bit more suited to this kind of thing?

IE JSON, YAML (used alot on the team), TOML, etc.

SuperZooper3 replied:
Personal preference and lack and needing any more libraries or code to process the config info. I can just import in the file. It also lets me put some sanitation logic at the bottom of the file, without having to break that off. But yeah, I'm not apposed to changing it if there is a reason to, lmk what you think @KavinSatheeskumar.

Darth-Raazi replied:
I agree with Kavin on this, Python files shouldn't be used for configs. I would opt for YAML in this case but feel free to use JSON or something else if you really want. This isn't super space-critical embedded code, so we don't really need to worry about not needing external libraries

KavinSatheeskumar reviewed (approved):
Hey, had a few minor points, but over all the code is fine. Feel free to address as you feel necessary

SuperZooper3 committed:
re-organised config file location

SuperZooper3 committed:
break running into a main.py

SuperZooper3 committed:
document start time offset

Darth-Raazi commented on main.py:
```diff
@@ -0,0 +1,3 @@
+from data_animation import run_animation
+
+run_animation()
```
Nitpick: End files with blank lines please

SuperZooper3 replied:
done

Darth-Raazi commented on data_animation.py:
```diff
@@ -0,0 +1,207 @@
+# It must then add all the data who's recorded timestamp is smaller than or equal to the
+# running time variable
+def animate(frame_number):
+    global time_cursor
+    global time
```
Oof using `global` should be avoided. Also to organize this code a bit more, can we refactor and put all the animation-related stuff in a class? That will solve both problems at once

Artem-Sotnikov replied:
Yeah I actually started doing a cleaning pass to address this issue. Stay tuned for a git push to clean this up.

SuperZooper3 replied:
Implemented (quite well may I say) by @Artem-Sotnikov in [856c997](https://github.com/waterloo-rocketry/data-video-visualization/pull/1/commits/856c99790e4501d3754d2777ea7af2ee0f3d0687).

Artem-Sotnikov commented on data_animation.py:
```diff
@@ -0,0 +1,207 @@
+        # Create the empty line object that will be filled out as the animiation runs
+        self.line = self.axis.plot([], [], label=f"{display_name}", color=color)[0]
+
+    """a tick pushing the data forward one recording frame, and this might happen multiple 
+    times per animation frame"""
```
This is a docstring for the function, which is an acceptable use case. In a perfect world we would have numpy/google type docstring for every function and class but that's excessive.

Commit:
Create rough but functional refactored version of data visualizer

Split up code among classes, add proper time estimation, add more error checking, add a few more comments. Code is functional but still need documentation in the readme + cleaning up of whitespace/newlines/line-length

Commit:
Clean up repo

Clean up line-lenght & whitespace, clean out extraneous files & organize repo structure

SuperZooper3 committed:
minor formatting

SuperZooper3 committed:
fix config example ffmpeg path and null

SuperZooper3 committed:
end main with newline

SuperZooper3 committed:
moved docstrings to under definition

Darth-Raazi commented on data_animation.py:
```diff
@@ -0,0 +1,269 @@
+        """a tick pushing the data forward one recording frame, and this might happen multiple 
+        times per animation frame.
+        """
+        # print(plotted_time)
```
Outdated comment

Darth-Raazi commented on data_animation.py:
```diff
@@ -0,0 +1,269 @@
+        self.time_cursor = 0
+        self.time = self.config_ani.STARTING_TIME  # animiation progress in seconds
+
+        self.fig, _ = plt.subplots()
```
Better practice to use `self.fig = plt.subplots()[0]`. `_` kills readability especially if it's used a lot

SuperZooper3 replied:
done

Darth-Raazi commented on .gitignore:
```diff
@@ -127,3 +127,15 @@ dmypy.json
+/working_directory 
+
+/config/*
+!config/config_ani_example.py
+!config/config_setup_example.py
```
These can be removed now

Darth-Raazi commented on working_directory/.gitignore:
```diff
@@ -0,0 +1,6 @@
+*
+!.gitignore
+
+# Trick for keeping empty folder in repository courtesy of
+# https://stackoverflow.com/questions/4250063/how-to-gitignore-all-files-folder-in-a-folder-but-not-the-folder-itself
```
Why do we need to have an empty folder on GH? Would it be better to have the user create one when needed (similar to how most C code compilations require you to create a `build` dir) and have the whole thing gitignored? Could maybe include a check at the beginning of the script to see if the directory exists and if not raise an error

SuperZooper3 replied:
Yeah, that's totally fair. This was just easier than writing the few lines to store an empty folder

Artem-Sotnikov replied:
This is a style thing but I don't think that the empty folder approach is entirely without merit. We're not dealing with a super complicated directory so just having an empty folder in place makes it easier on the user and more idiot-proof. We can go without here but that's my five cents on the matter.

Darth-Raazi reviewed (changes requested):
Great work with the refactor, just a few small things I noticed

SuperZooper3 committed:
remove working directory from tracking, add check

SuperZooper3 committed:
minor formatting

SuperZooper3 committed:
Correct yaml formatting

SuperZooper3 commented on config/config_ani_example.yaml:
```diff
@@ -0,0 +1,71 @@
+        "max": 30,
+        "primary": False
+    }
+}
```
done

Darth-Raazi reviewed (approved):
LGTM, great work guys
